### PR TITLE
Restructure the project for new versioning approach

### LINKED ISSUE #45: Decide on final versioning structure
**This issue currently block most of the API development of the project**

Originally, I intended this project to have relatively simple versioning, by simply splitting the packets folder into several version subdirectories like `v757`, holding all of the packets for that specific version. This was the intuitive way to handle this, that didn't really require much thought.

However after actually designing this out, it turned out that this kind of versioning brings a few issues in:

## Code duplication

Obviously, having version folders with the packets for corresponding minecraft versions will mean the content in these folders will unavoidably have to be duplicated. At the very least, this duplication would be structural, i.e. in the file structure, so each version would have say a folder for handshake state packets, login state, ...  and each of those folder would then have files containing the actual packets. This alone is pretty annoying, but it's not the worst thing about it.

As to the actual code in these files, there's 2 choices we can take:

<ul>
<li>
<details>
<summary><b>Copy-paste method</b></summary>

The obvious thing that comes to mind is to simply copy-paste the packets into version folders. However this means a complete code duplication (and therefore probably also test duplication), which comes with the common code duplication downsides, that is, if the code needs to change for whatever reason (say there's a bug), it now needs to change everywhere. 

This leaves you having to make the same change in potentially dozen of files, which are mostly the same, but might have slight differences that you potentially need to pay attention to. This gets super annoying very quickly, it's hard to review, and can lead to a lot of issues where you don't pay attention to the slight differences between the versions and miss something.

</details>
</li>

<li>
<details>
<summary><b>Inheritance method</b></summary>

This method reduces a lot of direct code duplication, by utilizing inheritance. Simply, we can just have all of the packet classes in a new version inherit from the packet classes from the previous version, and only override the methods that need changing (if any), to handle the differences between protocol versions. This would result in code like this:

```python
from mcproto.packets.v757.handshaking.handshake import Handshake as OldHandshake

class Handshake(OldHandshake):
   ...
```

On one hand, this avoids a lot of code duplication, but it brings in a whole set of other issues, related to this method quickly becoming a dependency hell:

<ul>
<li>
<details>
<summary>Performance</summary>

If all of the classes inherit from the previous version, the more versions we add, the less performant the new packets will be, as they'll have bigger and bigger MRO. I'm not sure just how big this performance decrease would be, but it's certainly there. It also becomes a pretty big import chain, as we're importing the class from previous version, which is doing the same from the previous, etc. Which means loading more code into memory. Overall, it's just not a great place to be in performance-wise.

</details>
</li>
<li>
<details>
<summary>Orientation in code</summary>

Another big issue is that such a dependency chain would make it very hard to actually know what's going on in a particular packet class. It would mean that you'd have to manually follow the dependency chain down to the class where the function you're looking for was last overwritten, and check at how it works. 

The madness comes when that class is making calls to more inherited internal functions, or utilizing `super` to call the previous version of that method. At that point, you need to keep in mind what this function is doing, follow the chain to find the super call, and put that together in your mind, and worse off, that function can also be making a super call.

It then becomes absolutely impossible to even just figure out what the code for that specific packet class would even look like when it's all put together, and very mentally straining to try and understand how the code works.

</details>
</li>
<li>
<details>
<summary>Typing issue: 1 `final` classes</summary>

Using inheritance like this actually brings in another issue, this time, a typing related one. The issue is that we're marking the packet classes as `@typing.final`, which makes the type-checker understand that this class is ready to be used, and needs to contain implementations for any abstract classes (coming from the `Packet` class ABC). This is really nice, as it avoids us forgetting to implement some method, and immediately results in typing errors when the Packet ABC would gain another abstract method, etc. However the problem is that final classes aren't meant to be inherited from. This means that any time we use a final class as a base for another class, we'd need a type ignore.

</details>
</li>
<li>
<details>
<summary>Typing issue 2: Breaking LSP</summary>

While the first typing issue isn't that huge, there's yet another typing issue, that's more fundamental. This one is about the ["Liskov Substitution Principle" / LSP](https://en.wikipedia.org//wiki/Liskov_substitution_principle). Essentially, LSP talks about how a class that inherits from some other parent class should keep all of the parent's methods fully compatible, so if a function takes an instance of class `X`, it can be passed class `Y` that inherits from `X` without issues.

However with this use-case, we'd be completely breaking this principle, since we can't guarantee that the child classes will be the same, due to potential unexpected protocol changes, that will need method overrides which might even have a different signature. What this comes down to is that the child packet classes won't actually be proper "childs", OOP-wise, and so this inheritance is fundamentally flawed. It would technically work, but it's really weird, and would likely need a lot of type-ignores.

</details>
</li>
<li>
<details>
<summary>Missleading isinstance checks</summary>

Along with this comes the issue of users potentially wanting to run an `isinstance` check, verifying that some packet they received is actually packet of type X. Issue now is that if the packet they're checking would be say a `Ping` packet from version 758, and they'd be checking if it's an instance of a v757 packet, this check would pass, because our new packets inherit from older versions. This may, or may not cause a lot of issues to the users.

</details>
</li>
</ul>

Clearly then, the inheritance method isn't actually a very good choice either.

</details>
</li>

</ul>

### Type classes

The minecraft protocol has some shared types between the packets, such as a type that holds NBT data, or a type holding Chat data. These, as anything in the protocol, are subject to change between different protocol versions. This means that we might need to introduce versioning not just for packets, but also for types. As an example, we can have `mcproto.types.v757.chat.Chat` class, representing the chat type.

<ul>
<li>
<details>
<summary><b>Issues with inheritance based packets</b></summary>

Unfortunately, versioning types causes even more typing trouble with inheritance based packet classes. The versioned packets would now need to use the appropriate versioned types, so a v757 packet should only use v757 type.

This would however mean that since each packet inherits from the previous one, the type in the previous one would not be compatible anymore. This means we'd need to modify the signature of every function to change the type annotation to now only accept the new version of the type class. That is for all functions which relied on this type, which will at the very least include every `__init__` for packets that used a custom type.

This alone breaks LSP for every such packet class that had it's type overwritten, meaning we'll need yet more type ignores, and it introduces yet more inconsistency in our class structure, as these parent classes make no sense type-wise, they're just shorthands to avoid code repetition.

</details>
</li>
<li>
<details>
<summary><b>Copy-paste method packets would technically work well though</b></summary>

If the packet classes each had their own rewrite, it would simply have to be adjusted everywhere, and it means these types will also have their own versioned rewrites for each version, meaning no typing issues at all.

</details>
</li>

</ul>

Then again, these types probably shouldn't really change between the versions, and 1 implementation might even be sufficient, but the issue is that they might change, and if that does happen, we'd have to release a completely breaking change, suddenly introducing versioning for the types.

### Connection classes??

What if something fundamentally changes between the protocol versions, and say writing varints becomes different, should we also version the connection classes? If so, it would mean having to maintain different, but very similar, codebases for all of the supported protocol versions, which I'm not a huge fan of.

While the way connection is handled is very unlikely to get changed, as it was like this for an incredibly long time, we can't actually be sure that it won't change...

---

## Ok, so what now?

Well, a decision on how to proceed needs to be made, because before that, any further progress on implementing new packets is sort of blocked, as most of the new packets will already require some custom types, and without knowing whether to version them or not, a lot of extra work might end up being spent on something that'll need rewriting anyway, due to a different decision about how to handle this being made here.

## Potential ways to handle this

<ul>
<li>
<details open>
<summary><b>Make the library only support a single (latest) version</b></summary>

This would be the simplest and most elegant solution. No code duplication, no inheritance hell, and very simple, clean and readable codebase, that simply goes with the flow as minecraft protocol changes.

However it isn't without issues:

<details>
<summary><b>Library Versioning</b></summary>

The library is currently using the semantic versioning model, which has 3 main version parts: major, micro, patch versions + optionally some extra optional info at the end. Semver standard states very specifically when should there be an update to which of these version components, that is, non-breaking bugfixes should be a patch version bump, new features should be a micro version bumps and only if there are some breaking changes should there be a major version bump.

It's safe to assume that any protocol version bump would come with a major version bump, as even if the protocol versions were essentially compatible (which does happen), we're officially dropping support for the old version, which qualifies as a breaking change.

However, what if we make a breaking change without a protocol version bump? Does that also mean a major version update? Semver says it should! However this kind of versioning could end up being pretty confusing to users, and it could be really hard to then understand what minecraft/protocol version is supported in say mcproto v8.4.2.

The alternative solution would be to use the minecraft protocol versions as the major versions, and even on a breaking change, we'd just change the minor version. This isn't great, as many package managers, like say `python-poetry` or `pipenv` don't expect breaking changes from a micro version update, and so users depending on the library would need to specify the dependency in a very specific way, to make these tools recognize when it's safe to update mcproto.

Another issue with choosing this versioning approach, with protocol version as the major component is that it would be very hard to move away from. Even now the protocol version is at v762, so if we wanted to move away, our next versions would need to have even bigger version numbers, which is super annoying, and could be very confusing as people might've gotten used to it being the protocol version.

</details>

<details>
<summary><b>Maintaining older versions?</b></summary>

If we do go this route, do we now need to consider whether to maintain older versions of mcproto, for previous protocol versions?  i.e. for mcproto `v757.5.2`, we could fix a bug and release `v757.5.3`, while focusing feature development on the newever `v758.X.Y` versions. Of course, this assumes we'd move to the protocol version based project versioning.

Without moving to this type of versioning, and staying at semantic versioning, we'd be facing a pretty big potential issue, that is, if mcproto v7.2.1 is for protocol version v757, and there's already mcproto v8.0.0 for v758, however a bug is found in the older version, that requires a breaking change, we can't just bump the major version as semver would tell us to, as 8.0.0 is already occupied.

</details>
 
</details>

<li>
<details open>
<summary><b>Use git branches for different versions, each branch only having support for it's version</b></summary>

- Essentially, this is just full code duplication, except this time it's not just code copied between files, it's between branches.
- How to ship this?

</details>
</li>

<li>
<details open>
<summary><b>Version the entire project, make <code>mcproto.v757</code>, <code>mcproto.v758</code>,... with very little being shared</b></summary>

This would essentially be full code duplication of huge amount of code, with the obvious issues coming with code duplication.

</details>
</li>

<li>
<details open>
<summary><b>Only version what needs to be versioned (packets, and perhaps types)</b></summary>

- What if something more basic changes? Like in the connection?
- Do we choose full code duplication in the individual versions, or inheritance (with typing issues)?

</details>
</li>

<li><b>Something I'm not thinking of</b></li>

</ul>

## My current opinion

I'm inclined towards just supporting a single version, and probably moving away from semantic versioning to a versioning scheme uses the packet version as the major component. We could probably even get away with versions consisting of 4 parts, protocol version + normal semver, not sure how well would tools like poetry play with this though, and it would mean there's pretty much no going back, as this would immediately make our major version a pretty big number, which would be very annoying if we go back and it could cause a lot of confusion.

As for maintaining older versions, I'd probably be open to bug fixes there, but I wouldn't carry features over, at least not in most cases. Also, these older versions wouldn't keep being supported forever, and there should be a time scheduler there.

### PR DESCRIPTION
**Pull request by ItsDrike**
This PR resolves #45, and restructures the entire project to no longer support multiple minecraft protocol versions, moving to a single (latest) version model.

Note that **this is a completely breaking change, and there will NOT be a deprecation period**. This is because of how this change impacts the rest of the project development. As this project is still in pre-release stage, deprecation here would greatly slow down any further development here, forcing contributors to maintain both the temporary deprecation layer, and the new single-version one.

For more details on how this change will affect the project going forward, and why it is necessary, I'd suggest checking out #45, containing a detailed explanation.

I have updated the README examples which now use this new version, so I'd suggest checking that out before moving forward with trying out this version, or reviewing the code, as it demonstrates how this new approach will work. Note that when testing this version out, it will almost certainly break any existing code-bases using mcproto, checking the README changes and looking at the changelog fragment message should make it clear on how to update.

---

Some decisions I made here, which I find potentially worth considering during a review. I'm open to doing these differently:

- The `generate_packet_map` function is dynamic - imports the appropriate libraries on function run, depending on the requested packets. The alternative here would be to simply hard-code packet maps dictionaries as constants. This could be done either in the new `packet_map.py` file, but also in the individual game state folders for the packets, such as `mcproto.packets.login import CLIENTBOUND_MAP`.
- The `generate_packet_map` uses caching, to avoid needlessly re-importing the modules and walking over all of the packets again. The clear downside of this is the fact that it uses more memory, though that's almost irrelevant as packet maps are relatively small.
- Because of how caching works, since we're returning a mutable dictionary  from `generate_packet_map`, the cache would hold the same reference as the returned dict, potentially causing issues if the user modifies that returned dict. ~~For this reason, I decided to also add another decorator, responsible for copying the result afterwards, making sure that the cache holds a different instance of the dictionary, and what users see is simply a copy of this instance, which they can modify or do whatever they wish with, without risking breaking future calls to this function.~~ For this reason, the function will now only return a mapping proxy, which is immutable, and holds a strong reference to the dict internally (as suggested by @Martysh12).

### CODE REVIEW

File: mcproto/packets/packet_map.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+
+class WalkableModuleData(NamedTuple):
+    module: ModuleType
+    info: pkgutil.ModuleInfo
+    member_names: Sequence[str]
+
+
+def _walk_submodules(module: ModuleType) -> Iterator[WalkableModuleData]:
```
[the remark this answers is not in the document]
Comment (ItsDrike) [reply mcproto/packets/packet_map.py:28]: I don't think this is a huge issue, it's an internal function, and the docstring describes what's being done here, I also don't really see a good way to make this function much simpler.

File: mcproto/packets/packet_map.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+@overload
+def generate_packet_map(
+    direction: Literal[PacketDirection.CLIENTBOUND],
+    state: GameState,
+) -> Mapping[int, type[ClientBoundPacket]]:
+    ...
+
+
+@lru_cache()
+def generate_packet_map(direction: PacketDirection, state: GameState) -> Mapping[int, type[Packet]]:
```
[the remark this answers is not in the document]
Comment (ItsDrike) [reply mcproto/packets/packet_map.py:98]: This function actually really isn't particularly complex or hard to understand.

**Review by Martysh12 — CHANGES_REQUESTED** [review]
That looks good to me, but that last point bugs me a little.
- I like the idea of importing packets dynamically, as maintaining a big dictionary of packets does seem like a pain.
- I don't mind the memory overhead caching gives the library, because importing is a slow procedure.
- Is it possible to return some kind of "immutable dictionary", so that you don't have to copy the dictionary? Maybe `MappingProxyType` from the `types` module? (available since Python 3.3)

**Comment by ItsDrike** [comment]
> Is it possible to return some kind of "immutable dictionary", so that you don't have to copy the dictionary? Maybe MappingProxyType from the types module? (available since Python 3.3)

This is a nice idea, however you need the real dictionary to be kept somewhere, and since a function call creates a new stackframe, with all of the variables in it, and as the function ends, that stackframe gets deleted, along with all the variables it held. You could probably do something hacky, like crating a global variable, holding a list of dictionaries, which are the return values, so that the caching then returns that mapping proxy back, with the actual dict being in that global variable list. But this is probably way too convoluted and I'm not sure I love that.

I do also find this caching with copying pretty annoying though, I just wasn't able to think of anything better. Btw, do you think the name `utils/decorators.py` works well here? Or can you think of a better name? I didn't want an overly specific name just for this copying logic, to avoid too many files in utils dir, but decorators might be a bit too broad. Not sure where else to put it / what to call that file.

**Comment by ItsDrike** [comment]
Hmm, actually after some testing, it seems like this would work, as `MappingProxyType` does count as an object holding a direct reference to the original `dict`, so it's more like returning a class holding a dict inside of it. I assumed mapping proxy worked more like a weakref. In this case I would agree that just returning a mapping proxy would probably work pretty well.

**Review by Martysh12 — APPROVED** [review]
> This is a nice idea, however you need the real dictionary to be kept somewhere, and since a function call creates a new stackframe, with all of the variables in it, as the function ends, that stackframe gets deleted, and that means taking all of the local variables it held alongside. A mapping proxy needs a reference to an actual dict object, so we can't have that dict going out of scope, while the mapping proxy is returned.
> You could probably do something hacky, like crating a global variable, holding a list of dictionaries, which are the return values, so that the caching then returns that mapping proxy back, with the actual dict being in that global variable list. But this is probably way too convoluted and I'm not sure I love that.

Riiight. Oh well, copying is not that bad after all.

> Btw, do you think the name `utils/decorators.py` works well here? Or can you think of a better name? I didn't want an overly specific name just for this copying logic, to avoid too many files in utils dir, but decorators might be a bit too broad. Not sure where else to put it / what to call that file.

I can't think of a better name than `utils/decorators.py`, so I guess it's final.

I'll approve the changes, then.

**Pull request merged**